src/sentry/replays/usecases/reader.py

```diff
@@ -0,0 +1,256 @@
+    return []
+
+
+def fetch_direct_storage_segment_meta(
+    project_id: int,
+    replay_id: str,
+    segment_id: int,
+) -> RecordingSegmentStorageMeta | None:
+    """Return direct-storage metadata derived from our Clickhouse table."""
+    if has_archived_segment(project_id, replay_id):
+        return None
+
+    results = _fetch_segments_from_snuba(
+        project_id=project_id,
+        replay_id=replay_id,
+        offset=0,
+        limit=1,
+        segment_id=segment_id,
+    )
+    if len(results) == 0:
+        return None
+    else:
+        return results[0]
+
+
+def has_archived_segment(project_id: int, replay_id: str) -> bool:
+    """Return true if an archive row exists for this replay."""
+    snuba_request = Request(
+        dataset="replays",
+        app_id="replay-backend-web",
+        query=Query(
+            match=Entity("replays"),
+            select=[Column("is_archived")],
+            where=[
+                Condition(Column("is_archived"), Op.EQ, 1),
+                Condition(Column("project_id"), Op.EQ, project_id),
+                Condition(Column("replay_id"), Op.EQ, str(uuid.UUID(replay_id))),
+                # We request the full 90 day range. This is effectively an unbounded timestamp
+                # range.
+                Condition(Column("timestamp"), Op.LT, datetime.now()),
+                Condition(Column("timestamp"), Op.GTE, datetime.now() - timedelta(days=90)),
+            ],
+            granularity=Granularity(3600),
+        ),
+    )
+    response = raw_snql_query(snuba_request, "replays.query.download_replay_segments")
+    return len(response["data"]) > 0
+
+
+def _fetch_segments_from_snuba(
+    project_id: int,
+    replay_id: str,
+    offset: int,
+    limit: int,
+    segment_id: int | None = None,
+) -> List[RecordingSegmentStorageMeta]:
+    conditions = []
+    if segment_id:
+        conditions.append(Condition(Column("segment_id"), Op.EQ, segment_id))
+
+    snuba_request = Request(
+        dataset="replays",
+        app_id="replay-backend-web",
+        query=Query(
+            match=Entity("replays"),
+            select=[Column("segment_id"), Column("retention_days"), Column("timestamp")],
+            where=[
+                Condition(Column("project_id"), Op.EQ, project_id),
+                Condition(Column("replay_id"), Op.EQ, str(uuid.UUID(replay_id))),
+                # We request the full 90 day range. This is effectively an unbounded timestamp
+                # range.
+                Condition(Column("timestamp"), Op.LT, datetime.now()),
+                Condition(Column("timestamp"), Op.GTE, datetime.now() - timedelta(days=90)),
+                # NOTE: Optimization to reduce the number of rows before the LIMIT clause. The
+                # cursors happen to map 1 to 1 with segment_id. If these filters are removed
+                # you will need to supply an offset value.
+                Condition(Column("segment_id"), Op.GTE, offset),
```

Comment:
if missing segments occur does this interfere with pagination in any way? worth adding any tests?

Reply:
Dependent on client behavior.  If it checks `len(segments)` and makes a pagination decision based on `< LIMIT`.  I think its safer to remove this and re-add if we need it.